R/module_nested_tabs.R

```diff
@@ -236,83 +258,83 @@ srv_nested_tabs.teal_module <- function(id, datasets, modules, is_module_specifi
-#' @param module (`teal_module`) module where needed filters are taken from
-#' @param datasets (`FilteredData`) object where needed data are taken from
-#'
-#' @return A `teal_data` object.
-#'
-#' @keywords internal
-.datasets_to_data <- function(module, datasets) {
-  checkmate::assert_class(module, "teal_module")
-  checkmate::assert_class(datasets, "FilteredData")
-
-  datanames <- if (is.null(module$datanames) || identical(module$datanames, "all")) {
-    datasets$datanames()
-  } else {
-    include_parent_datanames(
-      module$datanames,
-      datasets$get_join_keys()
+.make_teal_data <- function(modules, data, datasets, datanames) {
+  new_datasets <- c(
+    # Filtered data
+    sapply(datanames, function(x) datasets$get_data(x, filtered = TRUE), simplify = FALSE),
+    # Raw (unfiltered data)
+    stats::setNames(
+      lapply(datanames, function(x) datasets$get_data(x, filtered = FALSE)),
+      sprintf("%s_raw", datanames)
     )
-  }
-
-  # list of reactive filtered data
-  data <- sapply(datanames, function(x) datasets$get_data(x, filtered = TRUE), simplify = FALSE)
-
-  hashes <- calculate_hashes(datanames, datasets)
-
-  code <- c(
-    get_rcode_str_install(),
-    get_rcode_libraries(),
-    get_datasets_code(datanames, datasets, hashes)
   )
 
+  data_code <- teal.data::get_code(data, datanames = datanames)
+  hashes_code <- .get_hashes_code(datasets, datanames)
+  raw_data_code <- sprintf("%1$s_raw <- %1$s", datanames)
+  filter_code <- get_filter_expr(datasets, datanames = datanames)
 
-  data <- do.call(
+  all_code <- paste(unlist(c(data_code, "", hashes_code, raw_data_code, "", filter_code)), collapse = "\n")
+  tdata <- do.call(
     teal.data::teal_data,
-    args = c(data, code = list(code), join_keys = list(datasets$get_join_keys()[datanames]))
+    c(
+      list(code = trimws(all_code, which = "right")),
+      list(join_keys = teal.data::join_keys(data)),
+      new_datasets
+    )
   )
 
-  data@verified <- attr(datasets, "verification_status")
-  data
+  tdata@verified <- data@verified
+  teal.data::datanames(tdata) <- datanames
+  tdata
 }
 
-#' Get the hash of a dataset
+#' Get code that tests the integrity of the reproducible data
 #'
-#' @param datanames (`character`) names of datasets
 #' @param datasets (`FilteredData`) object holding the data
+#' @param datanames (`character`) names of datasets
 #'
-#' @return A list of hashes per dataset.
+#' @return A character vector with the code lines.
 #' @keywords internal
 #'
-calculate_hashes <- function(datanames, datasets) {
-  sapply(datanames, function(x) rlang::hash(datasets$get_data(x, filtered = FALSE)), simplify = FALSE)
+.get_hashes_code <- function(datasets, datanames) {
+  vapply(
+    datanames,
+    function(dataname, datasets) {
+      hash <- rlang::hash(datasets$get_data(dataname, filtered = FALSE))
+      sprintf(
+        "stopifnot(%s == %s)",
+        deparse1(bquote(rlang::hash(.(as.name(dataname))))),
+        deparse1(hash)
+      )
```

Comment:
either two helper vars or two nested calls
```suggestion
      sprintf(
        "stopifnot(%s == %s)",
        deparse1(bquote(rlang::hash(.(as.name(dataname))))),
        deparse1(rlang::hash(datasets$get_data(dataname, filtered = FALSE)))
      )
```